### PR TITLE
fixme: delete toast values of invisible AppendOptimized tuple during vacuum.

### PR DESCRIPTION
**Pull request by QingMa-gp**
This PR contains the following changes:  
1. Resolves GPDB_12_MERGE_FIXME in  appendonly_compaction.c:

    > GPDB_12_MERGE_FIXME: loop through all attributes, call `toast_delete_datum()`
    > on any toasted datums, like `toast_delete` does for heap tuples.

2. Refactor `ExecFetchSlotMemTuple()`：
     - Rename `ExecFetchSlotMemTuple()` to `appendonly_form_memtuple()` like `heap_from_tuple()`.
     - Remove `shouldFree` parameter since it's always set as true.
     - Add function `appendonly_free_memtuple()` to replace `free(memtuple)`.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by kainwen — COMMENTED** [review]
Code change LGTM.

Will the result number in test output be stable? Will they be impacted by other cases?

**Comment by QingMa-gp** [comment]
Thanks for your review. @kainwen 
> Will the result number in test output be stable? 

Yes. It's not stable because there is random text on the distribution column. It has been fixed.
> Will they be impacted by other cases?

I don't think it will be affected by other cases, because `TOAST_MAX_CHUNK_SIZE` is only affected by `BLCKSZ`, which is fixed. Am I right about this?

File: src/backend/access/appendonly/appendonly_compaction.c [hunk 1]
```diff
@@ -319,31 +319,53 @@ AppendOnlyMoveTuple(TupleTableSlot *slot,
...
+	aoTupleId = (AOTupleId *) &slot->tts_tid;
 	/* Extract all the values of the tuple */
 	slot_getallattrs(slot);
 
-	/* GPDB_12_MERGE_FIXME: loop through all attributes, call toast_delete_datum()
-	 * on any toasted datums, like toast_delete does for heap tuples */
-#if 0
-	tuple = TupGetMemTuple(slot);
-	if (MemTupleHasExternal(tuple, mt_bind))
+	tuple = ExecFetchSlotMemTuple(slot, &shouldFree, mt_bind);
```
Comment (haolinw) [inline src/backend/access/appendonly/appendonly_compaction.c:339]: The function name "ExecFetchSlotMemTuple" is a little bit confused that it's actually not the executor layer API but with "Exec" prefix.

Some suggestion on that:
- Changing the name to appendonly_form_memtuple or so.
- Remove shouldFree parameter, and the following if(shouldFree) pfree... clause, replace it with appendonly_free_memtuple() (internal call pfree) like heap_freetuple.
- Apply above to all place where calls ExecFetchSlotMemTuple().

File: src/backend/access/appendonly/appendonlyam_handler.c [hunk 2]
```diff
@@ -371,6 +369,15 @@ ExecFetchSlotMemTuple(TupleTableSlot *slot, bool *shouldFree, MemTupleBinding *m
 	return result;
 }
 
+/*
+ * appendonly_free_memtuple
+ */
+void
```
[the remark this answers is not in the document]
Comment (haolinw) [reply src/backend/access/appendonly/appendonlyam_handler.c:375]: Just want to keep same convention as PG like pattern form_tuple/freetuple pair, maybe better to remind developer to free the tuple after using, of course form_tuple does not have strong semantic of "alloc memory" as malloc or alloc, a careless developer might forget calling free. If we can rename it better like appendonly_alloc_tuple ? We can ignore the wrapper of the free.

**Comment by haolinw** [comment]
LGTM

**Review by haolinw — APPROVED** [review]

**Comment by QingMa-gp** [comment]
Thanks for the review.

**Pull request merged**